File: dotnet/src/Connection/Unregister.cs
namespace Pong
{
  partial class Connection
  {
    /// <summary>
    /// A message that is sent when a user unregisters
    /// <br/>
    /// Sent to all players who aren't in a session
    /// </summary>
    public class UnregisterEvent : GameEvent
    {
      public string Name { get; set; }

      public UnregisterEvent(string name)
        : base("unregister")
      {
        Name = name;
      }

      /// <summary>
      /// Deserialize an UnregisterEvent from json
      /// </summary>
      public static new UnregisterEvent Deserialize(string json)
      {
        return Deserialize<UnregisterEvent>(json);
      }
    }

    /// <summary>
    /// Unregister with the server
    /// </summary>
    public async Task Unregister()
    {
      var UnregisterEvent = new Dictionary<string, string> { { "type", "unregister" } };

      await Send(UnregisterEvent);
    }
  }
}

File: server/src/messages/unregister.ts
import Player from "@/players/player";

/**
 * A message that is sent when a user unregisters
 * Sent to all players who aren't in a session
 */
export interface UnregisterMessage {
  type: "unregister";
  name: string;
}

/**
 * Unregister a user
 * Sent to all players who aren't in a session
 * @param {Player} player The player who registered
 * @returns {UnregisterMessage} A Register message
 */
export function Unregister(player: Player): UnregisterMessage {
  return {
    type: "unregister",
    name: player.name,
  };
}

export default Unregister;
